### ISSUE TITLE
Exported TypeScript namespaces don’t compile properly with new SWC compiler

### ISSUE THREAD

**Issue by GrantASL19** [vercel/next.js#30379]
### What version of Next.js are you using?

12.0.0

### What version of Node.js are you using?

16.6.2

### What browser are you using?

N/A

### What operating system are you using?

macOS

### How are you deploying your application?

N/A

### Describe the Bug

In our Next.js 11 sites we use exported TypeScript namespaces for some shared package and style modules. It’s maybe ill-advised to use namespaces like this but it lets us write code like this:

```ts
namespace appValues {
  const firstName = "Foo";
  const lastName = "Bar";

  export const fullName = `${firstName} ${lastName}`
}

export default appValues;
```

Which is roughly equivalent to this:

```ts
const firstName = "Foo";
const lastName = "Bar";

const appValues = {
  fullName: `${firstName} ${lastName}`
}

export default appValues;
```

Unfortunately this isn’t working in Next.js 12 when SWC is used, presumably because SWC compiles this namespace code (AFAIK) incorrectly.

### Expected Behavior

If `appValues` is imported in another file this code should log "Foo Bar":

```ts
import appValues from 'appValues'

console.log(appValues.fullName)
```

This would work in Next.js 11, but fails in Next.js 12 with `Cannot read property 'fullName' of undefined`.

### To Reproduce

Here’s a reproduction based on the latest `create-next-app` (with the `typescript` preset):

https://github.com/GrantASL19/nextjs-12-typescript-namespace-default-export-error

Note that VS Code doesn’t flag any type errors, but the build fails.

**Comment by hockdudu** [comment]
I also have the same problem on my project after updating it to Next.js 12.

```typescript
namespace DateFormatter {
   // ...
}

export default DateFormatter;
```

I noticed that if I save the namespace to a variable first, it then works

**Comment by hockdudu** [comment]
I have the same problem on my project after updating to Next.js 12.

```typescript
namespace DateFormatter {
    export function getFormattedDate(date: Date): string {
        // ...
    }
}

export default DateFormatter;
```

If I try to use it somewhere else:
```typescript
const formattedDate = DateFormatter.getFormattedDate(new Date());
```

`TypeError: Cannot read properties of undefined (reading 'getFormattedDate')`

---

I noticed, if I save the namespace to a variable first, then it works:

```typescript
namespace DateFormatter {
    // ...
}

const dateFormatter = DateFormatter;

export default dateFormatter;
```

Nonetheless, I should be able to export the namespace without saving it to a variable first.

**Comment by anvelicon** [comment]
I am encountering the same issue while exporting an object.

When doing something like this:
```typescript
import someObject from '../lib/someObject'

const {result} = someObject.attribute()
```

This seems to be the behaviour:
```typescript
/* 
 This does not work
*/
const someObject = {
 attribute: () => {
 // Instructions
 return { result }
 }
}

export default someObject


/* 
 This does
*/
const anotherVariable = () => {
 // Instructions
 return { result }
}

const someObject = {
 attribute: anotherVariable
}

export default someObject
```
The error i get in the first case is:
`TypeError: Cannot read properties of undefined (reading 'result')`

**Comment by Sheldonfrith** [comment]
I'm having a similar issue, probably related, but there is no wrapping object, it seems just the whole module is undefined or something. And it does seem to be during the compilation stage as well for me.

Error: ```TypeError: Cannot read properties of undefined (reading 'getSomething')" ... in 'file-2.ts' ```

###file-1.ts
```
export function getSomething(someArgs){
     //includes references to other functions and other modules
}
```

###file-2.ts
```
import { getSomething } from "./file-1";

getSomething(...);
```

**Comment by GrantASL19** [comment]
FWIW I updated [my reproduction](https://github.com/GrantASL19/nextjs-12-typescript-namespace-default-export-error) to Next.js `12.1.1-canary.14` and this is still a problem.

Now that the [Emotion transform](https://nextjs.org/docs/advanced-features/compiler#emotion) is close this namespace issue is the only blocker for us removing our Babel config.

**Comment by GrantASL19** [comment]
I [updated my reproduction again](https://github.com/GrantASL19/nextjs-12-typescript-namespace-default-export-error/commit/7099535c0ac95ca7562466d1e8a40c8e9c3f3577).

In case anyone else is blocked by this, I also found a simpler fix, which I’ve included in a comment in my repo. I hope this also helps debug the issue in SWC/Next.js.

If this module is exported from another module it will be `undefined`:

```ts
namespace appValues {
  const firstName = "Foo";
  const lastName = "Bar";
  export const fullName = `${firstName} ${lastName}`
}

export default appValues;
```

Whereas this will work as expected (replaced `export default appValues;` with `export default (appValues);`:

```ts
namespace appValues {
  const firstName = "Foo";
  const lastName = "Bar";
  export const fullName = `${firstName} ${lastName}`
}

export default (appValues);
```

**Comment by balazsorban44** [comment]
This issue has been automatically closed because it wasn't verified against next@canary. If you think it was closed by accident, please leave a comment. If you are running into a similar issue, please open a new issue with a reproduction. Thank you.
